File: prev-2024/3005.count-elements-with-maximum-frequency.js
/*
 * @lc app=leetcode id=2971 lang=javascript
 *
 * [3005] Count Elements With Maximum Frequency
 */
// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxFrequencyElements = function (nums) {
  /** @type {Map<number, number>} */
  const numToFreq = new Map();
  /** @type {Map<number, Set<number>>} */
  const freqToNums = new Map();

  for (const num of nums) {
    const freq = numToFreq.get(num) ?? 0;
    const freqNums = freqToNums.get(freq);
    freqNums?.delete(num);
    const nextFreqNums = freqToNums.get(freq + 1) ?? new Set();
    nextFreqNums.add(num);
    freqToNums.set(freq + 1, nextFreqNums);
    numToFreq.set(num, freq + 1);
  }

  const maxFreq = freqToNums.size;
  return maxFreq * freqToNums.get(maxFreq)?.size;
};
// @lc code=end

File: src/2025/3005.count-elements-with-maximum-frequency.cpp
/*
 * @lc app=leetcode id=3005 lang=cpp
 *
 * [3005] Count Elements With Maximum Frequency
 */

#include <algorithm>
#include <vector>

// @lc code=start
class Solution {
 public:
  int maxFrequencyElements(const std::vector<int>& nums) {
    std::array<int, 100> freqs = {};
    int max_freq = 0;
    int max_freq_element_count = 0;
    for (const int num : nums) {
      freqs[num - 1]++;
      if (freqs[num - 1] > max_freq) {
        max_freq_element_count = 1;
        max_freq = freqs[num - 1];
      } else if (freqs[num - 1] == max_freq) {
        max_freq_element_count++;
      }
    }
    return max_freq_element_count * max_freq;
  }
};
// @lc code=end
